cellxgene_schema_cli/cellxgene_schema/atac_seq.py

```diff
@@ -0,0 +1,355 @@
+    if unique_organism_ontology_term_ids != unique_feature_reference:
+        return "Unique Anndata.obs.organism_ontology_term_id must be equal to unqiue Anndata.var.feature_reference."
+
+
+def detect_chromosomes(parquet_file: str) -> list[str]:
+    logger.info("detecting chromosomes")
+    df = ddf.read_parquet(parquet_file, columns=["chromosome"]).drop_duplicates()
+    return df["chromosome"].values.compute()
+
+
+def index_fragment(fragment_file: str, parquet_file: str, tempdir: str):
+    # sort the fragment by chromosome, start coordinate, and stop coordinate, then compress it with bgzip
+    bgzip_output_file = fragment_file.replace(".gz", ".bgz")
+    bgzip_output_path = Path(bgzip_output_file)
+    bgzip_output_path.unlink(missing_ok=True)
+    bgzip_output_path.touch()
+    bgzip_write_lock = dd.Lock()  # lock to preserver write order
+
+    if not shutil.which("bgzip"):  # check if bgzip cli is installed
+        logger.warning("bgzip is not installed, using slower pysam implementation")
+        write_algorithm = write_algorithm_by_callable["pysam"]
+    else:
+        write_algorithm = write_algorithm_by_callable["cli"]
+
+    chromosomes = detect_chromosomes(parquet_file)
+    jobs = prepare_fragment(chromosomes, parquet_file, bgzip_output_file, tempdir, bgzip_write_lock, write_algorithm)
+    # limit calls to dask.compute to improve performace. The number of jobs to run at once is determined by the
+    # step variable. If we run all the jobs in the same call to dask.compute, the local cluster hangs.
+    # TODO: investigate why
+    step = 4
+    # print the progress of the jobs
+    for i in range(0, len(jobs), step):
+        dask.compute(jobs[i : i + step])
+
+    logger.info(f"Fragment sorted and compressed: {bgzip_output_file}")
+
+    pysam.tabix_index(bgzip_output_file, preset="bed", force=True)
+    tabix_output_file = bgzip_output_file + ".tbi"
+    logger.info(f"Index file generated: {tabix_output_file}")
+
+
+@delayed
+def sort_fragment(parquet_file: str, write_path: str, chromosome: str) -> Path:
+    temp_data = Path(write_path) / f"temp_{chromosome}.tsv"
+    df = ddf.read_parquet(parquet_file, filters=[("chromosome", "==", chromosome)])
+    df = df[column_ordering]
+    df = df.sort_values(["start coordinate", "stop coordinate"])
+
+    df.to_csv(temp_data, sep="\t", index=False, header=False, mode="w", single_file=True)
+    return temp_data
+
+
+@delayed
+def write_bgzip_pysam(input_file: str, bgzip_output_file: str, write_lock: dd.Lock):
+    with write_lock, pysam.libcbgzf.BGZFile(bgzip_output_file, mode="ab") as f_out, open(input_file, "rb") as f_in:
+        while data := f_in.read(2**20):
+            f_out.write(data)
+
+
+@delayed
+def write_bgzip_cli(input_file: str, bgzip_output_file: str, write_lock: dd.Lock):
+    with write_lock:
+        subprocess.run([f"cat {input_file} | bgzip --threads=8 -c >> {bgzip_output_file}"], shell=True, check=True)
```

Comment:
The current implementation could be vulnerable to [injection attacks](https://docs.python.org/3/library/subprocess.html#security-considerations). This should fix that:

```suggestion
    with write_lock, open(input_file, "rb") as fin, open(output_file, "ab") as fout:
        subprocess.run(
            ["bgzip", "--threads=8", "-c"], 
            stdin=fin, 
            stdout=fout, 
            check=True
        )
```